bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,595 @@
+    private static final String HUB_REST_REQUEST_MACADDRESS = "{\"ID\":%d,\"Service\":\"SystemInfo\"}\0";
+    private static final String HUB_TOKEN_SERVICE = "Service";
+    private static final String HUB_TOKEN_ZID = "ZID";
+    private static final String HUB_TOKEN_PROPERTY_LIST = "PropertyList";
+    private static final String HUB_TOKEN_DEVICE_TYPE = "DeviceType";
+    private static final String HUB_TOKEN_SWITCH = "Switch";
+    private static final String HUB_TOKEN_DIMMER = "Dimmer";
+    private static final String HUB_TOKEN_NAME = "Name";
+    private static final String HUB_TOKEN_POWER = "Power";
+    private static final String HUB_TOKEN_POWER_LEVEL = "PowerLevel";
+    private static final String HUB_TOKEN_MAC_ADDRESS = "MACAddress";
+    private static final String HUB_TOKEN_ZONE_LIST = "ZoneList";
+    private static final String HUB_SERVICE_REPORT_ZONE_PROPERTIES = "ReportZoneProperties";
+    private static final String HUB_SERVICE_ZONE_PROPERTIES_CHANGED = "ZonePropertiesChanged";
+    private static final String HUB_SERVICE_LIST_ZONE = "ListZones";
+    private static final String HUB_SERVICE_SYSTEM_INFO = "SystemInfo";
+
+    private @Nullable Future<?> hubController;
+    private String hubHost;
+    private int hubPort;
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable InputStreamReader hubInReader;
+    private @Nullable JsonStreamParser hubIn;
+    private CompletableFuture<@Nullable Void> hubControllerConnected;
+    private int hubReconnectSleep; // Sleep time before we attempt re-connect
+    private ScheduledExecutorService scheduler;
+
+    private volatile boolean stop; // Stop the controller as soon as possible
+    private volatile boolean stopWhenCommandsServed; // Stop the controller once all pending commands have been served
+    private volatile long stopTimestamp = 0; // Stop the controller after this timestamp
+
+    // When we submit commmands to the hub we don't correlate commands and responses. We simply use the first available
+    // response that answers our question. For that we maintain lists of all pending commands.
+    private Map<Integer, CompletableFuture<AdorneDeviceState>> stateCommands;
+    private @Nullable CompletableFuture<List<Integer>> zoneCommand;
+    private @Nullable CompletableFuture<String> macAddressCommand;
+    private int commandId; // We assign increasing command ids to all rest commands to the hub for easier
+                           // troubleshooting
+
+    private volatile @Nullable AdorneHubChangeNotify changeListener;
+
+    private final Object stopLock = new Object();
+    private final Object hubOutLock = new Object();
+    private final Object stateCommandsLock = new Object();
+    private final Object macAddressCommandLock = new Object();
+    private final Object zoneCommandLock = new Object();
+
+    public AdorneHubController(AdorneHubConfiguration config, ScheduledExecutorService scheduler) {
```

Comment:
Why don't you include `changeListener` as a construction parameter? That way you can make it non-null and final.

Reply:
It seems a bit heavy handed to require a listener, but it certainly makes the code simpler (also removed the setter). Drawback is that I had to implement a dummy listener for the discovery service, but not a big deal.